crates/sui-types/src/nitro_attestation.rs

```diff
@@ -391,138 +400,215 @@ impl AttestationDocument {
     /// Parse the payload of the attestation document, validate the cert based on timestamp, and the pcrs match.
     /// Adapted from <https://github.com/EternisAI/remote-attestation-verifier/blob/main/src/lib.rs>
     pub fn parse_payload(
-        payload: &Vec<u8>,
+        payload: &[u8],
     ) -> Result<AttestationDocument, NitroAttestationVerifyError> {
-        let document_data: ciborium::value::Value = ciborium::de::from_reader(payload.as_slice())
-            .map_err(|err| {
-            NitroAttestationVerifyError::InvalidAttestationDoc(format!(
-                "cannot parse payload CBOR: {}",
-                err
-            ))
-        })?;
+        let document_map = Self::to_map(payload)?;
+        Self::validate_document_map(&document_map)
+    }
+
+    fn to_map(payload: &[u8]) -> Result<HashMap<String, Value>, NitroAttestationVerifyError> {
+        let document_data: ciborium::value::Value =
+            ciborium::de::from_reader(payload).map_err(|err| {
+                NitroAttestationVerifyError::InvalidAttestationDoc(format!(
+                    "cannot parse payload CBOR: {}",
+                    err
+                ))
+            })?;
 
-        let document_map = match document_data {
-            ciborium::value::Value::Map(map) => map,
+        let document_map: HashMap<String, Value> = match document_data {
+            ciborium::value::Value::Map(map) => map
+                .into_iter()
+                .map(|(k, v)| {
+                    let k =
+                        k.as_text()
+                            .ok_or(NitroAttestationVerifyError::InvalidAttestationDoc(format!(
+                                "invalid key type: {:?}",
+                                k
+                            )))?;
+                    Ok((k.to_string(), v))
+                })
+                .collect::<Result<HashMap<String, Value>, NitroAttestationVerifyError>>()?,
             _ => {
                 return Err(NitroAttestationVerifyError::InvalidAttestationDoc(format!(
                     "expected map, got {:?}",
                     document_data
                 )))
             }
         };
+        Ok(document_map)
+    }
 
-        let get_string = |key: &str| -> Result<String, NitroAttestationVerifyError> {
-            match document_map
-                .iter()
-                .find(|(k, _)| matches!(k, ciborium::value::Value::Text(s) if s == key))
-            {
-                Some((_, ciborium::value::Value::Text(val))) => Ok(val.clone()),
-                _ => Err(NitroAttestationVerifyError::InvalidAttestationDoc(format!(
-                    "cannot parse {}",
-                    key
-                ))),
-            }
-        };
+    fn validate_document_map(
+        document_map: &HashMap<String, Value>,
+    ) -> Result<AttestationDocument, NitroAttestationVerifyError> {
+        let module_id = document_map
+            .get("module_id")
```

Comment:
just making sure it's the case - `.get()` is implemented as a map in O(1) query time, and not using a linear search.

Reply:
correct, hashmap is o(1) query